### ISSUE TITLE
Theme's Stylesheet May Be Getting Expanded Every Idle Period

### ISSUE THREAD

**Issue by tbpassin** [leo-editor/leo-editor#2452]
See PR #2456 and PR #2478.

I have been working on code to automatically pick up theme colors and recolor my output to match the applied theme.  While testing on all the themes in Leo's theme directory, I have noticed really strange behavior. It looks like the stylesheet is getting expanded either on most body keystrokes or (more likely) at every idle period.

This occurs using the "swotlight" theme.  This theme is very faulty (probably predates the current theme's logical names) because the following gets printed to the console multiple times when the stylesheet is constructed:

```
expand_css_constants Unresolved @constants
[
    '@log_text_background_color',
    '@log_text_foreground_color',
    '@text-foreground-bright',
    '@bg-gradient'
]
```

I am observing that this same message keeps getting printed over and over every few keystrokes in the body (but not when typing in a headline).  It's easy to miss after the screen fills up with those messages, but sometimes I can see the screen jump as it prints them over and over.

It is also happening when I select text in the body by dragging the mouse over the text.  Again, this only happens in the body but not the headline or minibuffer.

I seems to me that the only way this could happen is if the stylesheet gets expanded every few keystrokes, presumably at idle time.  This has to take a lot of computing cycles, and would tend to make Leo's responsiveness more sluggish.

With a better- constructed theme, these particular errors don't occur and so there is no message printed.  But presumably the same expansion will be happening unseen.

This behavior needs to be corrected.

**Comment by edreamleo** [comment]
@tbpassin Traces in expand_css_constants and ssm.do_pass do not show any activity at idle time or on keystrokes. It's possible that decluttering might trigger calls to expand_css_contants, but I don't think this is a serious issue.

**Comment by tbpassin** [comment]
I'm not using decluttering as far as I know.  The situation in which I saw the behavior was when the theme called for symbolic values (i.e., starting with "@") that were not defined elsewhere in the theme.  It only occurred for typing or mousing in the body, not the headline.

My concern is not whether a faulty theme is being handled correctly despite its error, but that this behavior might be going on invisibly when the theme has no error.

Would you repeat your tests using the swotLight theme, so we can see if they show something then?  That's the one theme in which I saw this issue.

**Comment by edreamleo** [comment]
@tbpassin Better yet, please insert g.trace(g.callers()) into the routines you suspect of being called too often.

**Comment by tbpassin** [comment]
This bug comes from the current line highlighter.  Every so often (probably during the idle period) the code tries to check whether the stylesheet colors have changed.  It yes, the new foreground and background colors are gotten are gotten from the stylesheet, which is expanded by `sheet = ssm.expand_css_constants(c1.active_stylesheet)`.

Apparently the check for changes doesn't work right and the stylesheet get expanded every time.  The expanded stylesheet can be gotten directly from the master widget, thus avoiding the expansion. I will put in a PR for this soon.

Longer term, the check for changes needs to be fixed.  I will work on that next.

**Comment by edreamleo** [comment]
Fixed via PR #2478.
